pwgen.1
.TH PWGEN 1 "August 2017" "pwgen version 2.08"
.SH NAME
pwgen \- generate pronounceable passwords
.SH SYNOPSIS
.B pwgen
[
.I OPTION
]
[
.I pw_length
]
[
.I num_pw
]
.SH DESCRIPTION
The
.B pwgen
program generates passwords which are designed to be easily memorized by
humans, while being as secure as possible.  Human-memorable passwords
are never going to be as secure as completely random
passwords.  In particular, passwords generated by
.B pwgen
without the 
.B \-s 
option should not be used in places where the password could be attacked
via an off-line brute-force attack.   On the other hand, completely
randomly generated  passwords have a tendency to be written down, 
and are subject to being compromised in that fashion.
.PP
The
.B pwgen
program is designed
to be used both interactively, and in shell scripts.  Hence, 
its default behavior differs depending on whether the standard output
is a tty device or a pipe to another program.  Used interactively, 
.B pwgen 
will display a screenful of passwords, allowing the user to pick a single 
password, and then quickly erase the screen.  This prevents someone from
being able to "shoulder surf" the user's chosen password.
.PP
When standard output (stdout) is not a tty, 
.B pwgen
will only generate one password, as this tends to be much more convenient
for shell scripts, and in order to be 
compatible with previous versions of this program.
.B
.SH OPTIONS
.TP
.B \-0, \--no-numerals
Don't include numbers in the generated passwords.
.TP
.B \-1
Print the generated passwords one per line.
.TP
.B \-A, \--no-capitalize
Don't bother to include any capital letters in the generated passwords.
.TP
.B \-a, --alt-phonics
This option doesn't do anything special; it is present only for
backwards compatibility.
.TP
.B \-B, --ambiguous
Don't use characters that could be confused by the user when printed,
such as 'l' and '1', or '0' or 'O'.  This reduces the number of possible 
passwords significantly, and as such reduces the quality of the
passwords.  It may be useful for users who have bad vision, but in
general use of this option is not recommended.
.TP
.B \-c, --capitalize
Include at least one capital letter in the password.  This is the default 
if the standard output is a tty device.
.TP
.B \-C
Print the generated passwords in columns.  This is the default if the 
standard output is a tty device.
.TP
.B \-e
Generate encrypted password strings alongside the cleartext passwords.
This defaults to sha256 ('$5$').  Use the environment variable
.I SALTPREFIX
to define the salt prefix ('', '$1$', '$5$', etc)
.TP
.B \-n, --numerals
Include at least one number in the password.  This is the default
if the standard output is a tty device.
.TP
.B \-N, --num-passwords=\fInum
Generate 
.I num
passwords.  This defaults to a screenful if passwords are 
printed by columns, and one password otherwise.
.TP
.B \-n, --numerals
Include at least one number in the password.  This is the default
if the standard output is a tty device.
.TP
.B \-H, --sha1=\fI/path/to/file[#seed]
Will use the sha1's hash of given file and the optional seed to create  
password. It will allow you to compute the same password later, 
if you remember the file, seed, and pwgen's options used.
ie: pwgen -H ~/your_favorite.mp3#your@email.com gives 
a list of possibles passwords for your pop3 account, and you can
ask this list again and again.
.IP
.B WARNING:
The passwords generated using this option are not very random.  If you use
this option, make sure the attacker can not obtain a copy of the file.
Also, note that the name of the file may be easily available from the
~/.history or ~/.bash_history file.
.TP
.B \-h, --help
Print a help message.
.TP
.B \-r \fIchars\fR, \fB--remove-chars=\fIchars
Don't use the specified characters in password.  This option will
disable the phomeme-based generator and uses the random password
generator.
.TP
.B \-s, --secure
Generate completely random, hard-to-memorize passwords.  These should
only be used for machine passwords, since otherwise it's almost
guaranteed that users will simply write the password on a piece of 
paper taped to the monitor...
.TP
.B \-v, --no-vowels
Generate random passwords that do not contain vowels or numbers that
might be mistaken for vowels.  It provides less secure passwords to
allow system administrators to not have to worry with random passwords
accidentally contain offensive substrings.
.TP
.B \-y, --symbols
Include at least one special character in the password. 
.SH AUTHOR
This version of 
.B pwgen
was written by Theodore Ts'o <tytso@alum.mit.edu>.  
It is modelled after a program
originally written by Brandon S. Allbery, and then
later extensively modified by Olaf Titz,  Jim Lynch, and others.
It was rewritten from scratch by Theodore Ts'o because the original program
was somewhat of a hack, and thus hard to maintain, and because 
the licensing status of the program was unclear.
.SH SEE ALSO
.BR passwd (1)


ChangeLog
2006-01-15  Theodore Ts'o  <tytso@mit.edu>

	* pwgen.1: Fix spelling mistakes in the man page (Addresses Debian
		Bug: #323538)

	* pwgen.c (main): Fix bug which would cause pwgen to loop forever
		if the password length specified is 1.

2005-06-15  Theodore Ts'o  <tytso@mit.edu>

	* pw_phonemes.c (pw_phonemes): Allow one or more capital letters
		and one or more digits in the generated passwords.
		(Addresses Debian Bug: #182595)

2005-06-14  Theodore Ts'o  <tytso@mit.edu>

	* pwgen.1: Explain that human-memorable passwords are vulnerable
		to off-line brute force attacks.  (Addresses Debian Bug:
		#276976)

	* pwgen.c, pwgen.h, pw_rand.c, pwgen.1: Make -A and -0 aliases to
		--no-capitalize and --no-numerals, and make those options
		work when --secure is specified.

	* pwgen.c, pwgen.h, pw_rand.c, pw_phonemes.c, pwgen.c: Add support
		for the --symbols/-y option, which adds special characters
		to the generated password.  (Addresses Debian Bug: #154561)
	
	* pwgen.c: Add -H, --sha1 option, to allow computing reproducible
		passwords, given a known file, and a known seed.
		(ie: pwgen -H ~/my_favourite.mp3#olivier@guerrier.com gives me
		a list of possibles passwords for my pop3 account, and I can
		ask this list again and again).   Feature suggested and
		implemented by Olivier Guerrier <olivier@guerrier.com>.

2005-06-13  Theodore Ts'o  <tytso@mit.edu>

	* Fix obvious spelling typo in pwgen.1.  (Addresses Debian Bug
		#311461)

2003-01-15  Theodore Ts'o  <tytso@mit.edu>

	* Release of pwgen 2.03

	* randnum.c: Fix #ifdef to use HAVE_DRAND48 instead of just
		RAND48.  This caused random() to be called without first
		initializing the right random number generator.  This
		apparently caused pwgen to core dump under AIX, which
		seems surprising.

	* pw_rand.c (pw_rand): Use a list of acceptable characters instead
		of a list of unacceptable characters.  This is slightly
		faster, and the code is cleaner and easier to understand.

	* randnum.c (get_random_fd): Use random() instead of rand() since
		we initialize the random number generator using srandom() if
		/dev/urandom and drand48() are not present.

	* Makefile.in (pwgen): Use LDFLAGS when linking pwgen.

	* pw_phonemes.c (pw_phonemes): Fix bug in how we call 
		pw_random_number() so that we generate passwords that
		start with vowels, and insert numbers from 0-9, not just
		0-8.  (Oops!)  Addresses Debian bug #176688.

2002-07-16  Theodore Ts'o  <tytso@mit.edu>

	* Makefile.in (install): Create the directory using $mandir, and
		not $(prefix)/man.

	* pw_phonemes.c: Add missing #include of string.h

2002-07-15  Theodore Ts'o  <tytso@mit.edu>

	* Release of pwgen 2.02

	* pwgen.c (main): If the number of characters is less than 5,
		force the use of the fully random generator, since the
		quality of phonetic generator is really bad if the lengths
		are small --- and if the length is less than or equal to
		2, pwgen will loop forever.  (Addresses Debian bug #117137)

	* pw_rand.c: Add most punctuation characters to the "bad
		characters list", so that the random passwords are much
		more usable.

	* pwgen.1: Fix e-mail address in man page.  (Addresses Debian bug
		#150579)

	* Makefile.in (WALL_OPTS): Add gcc options to do lots of nit-picky
		checking.

	* pwgen.c, pw_phonemes.c, pw_rand.c, pwgen.c, pwgen.h, randnum.c:
		Add const declarations and other code cleanups for
		avoidining -Wall warnings.

	* pwgen.c, pw_rand.c: #include the appropriate header files to
		avoid -Wall warnings.  (Addresses Debian bug #126774).

	* pwgen.c (main): Fix floating point exception errors if the
		password is longer than terminal width.  Addresses Debian
		bug #113571.

	* pwgen.c (main): Return an exit value of 0 when pwgen on
		success.  (Addresses Debian bugs #118181 and #102063).

2001-06-19  Theodore Tso  <tytso@valinux.com>

	* Release of pwgen 2.01

	* pwgen.c (usage): Add help/usage message



pwgen.c
/*
 * pwgen.c --- generate secure passwords
 *
 * Copyright (C) 2001,2002 by Theodore Ts'o
 * 
 * This file may be distributed under the terms of the GNU Public
 * License.
 */

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <crypt.h>
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif

#include "pwgen.h"

/* Globals variables */
int (*pw_number)(int max_num);

/* Program parameters set via getopt */

int	pw_length = 8;
int	num_pw = -1;
int	pwgen_flags = 0;
int	do_columns = 0;
int do_crypts = 0;
const char *salt_prefix = "$5$";

#ifdef HAVE_GETOPT_LONG
struct option pwgen_options[] = {
	{ "alt-phonics", no_argument, 0, 'a' },
	{ "capitalize", no_argument, 0, 'c' },
	{ "numerals", no_argument, 0, 'n'},
	{ "symbols", no_argument, 0, 'y'},
	{ "num-passwords", required_argument, 0, 'N'},
	{ "remove-chars", required_argument, 0, 'r' },
	{ "secure", no_argument, 0, 's' },
	{ "help", no_argument, 0, 'h'},
	{ "no-numerals", no_argument, 0, '0' },
	{ "no-capitalize", no_argument, 0, 'A' },
	{ "sha1", required_argument, 0, 'H' },
	{ "crypt", no_argument, 0, 'e' },
	{ "ambiguous", no_argument, 0, 'B' },
	{ "no-vowels", no_argument, 0, 'v' },
	{ 0, 0, 0, 0}
};
#endif

const char *pw_options = "01AaBCcnN:sr:hH:evy";

static void usage(void)
{
	fputs("Usage: pwgen [ OPTIONS ] [ pw_length ] [ num_pw ]\n\n", stderr);
	fputs("Options supported by pwgen:\n", stderr);
	fputs("  -c or --capitalize\n", stderr);
	fputs("\tInclude at least one capital letter in the password\n", 
	      stderr);
	fputs("  -A or --no-capitalize\n", stderr);
	fputs("\tDon't include capital letters in the password\n", 
	      stderr);
	fputs("  -n or --numerals\n", stderr);
	fputs("\tInclude at least one number in the password\n", stderr);
	fputs("  -0 or --no-numerals\n", stderr);
	fputs("\tDon't include numbers in the password\n", 
	      stderr);
	fputs("  -y or --symbols\n", stderr);
	fputs("\tInclude at least one special symbol in the password\n", 
	      stderr);
	fputs("  -r <chars> or --remove-chars=<chars>\n", stderr);
	fputs("\tRemove characters from the set of characters to "
	      "generate passwords\n", stderr);
	fputs("  -s or --secure\n", stderr);
	fputs("\tGenerate completely random passwords\n", stderr);
	fputs("  -B or --ambiguous\n", stderr);
	fputs("\tDon't include ambiguous characters in the password\n", 
	      stderr);
	fputs("  -h or --help\n", stderr);
	fputs("\tPrint a help message\n", stderr);
	fputs("  -H or --sha1=path/to/file[#seed]\n", stderr);
	fputs("\tUse sha1 hash of given file as a (not so) random generator\n",
	      stderr);
	fputs("  -C\n\tPrint the generated passwords in columns\n", stderr);
	fputs("  -1\n\tDon't print the generated passwords in columns\n", 
	      stderr);
	fputs("  -e\n\tPrint the encrypted crypt after password.\n", stderr);
	fputs("\tset salt prefix in SALTPREFIX environment ('', $1$, $5$, etc)\n", stderr);
	fputs("  -v or --no-vowels\n", stderr);
	fputs("\tDo not use any vowels so as to avoid accidental nasty words\n",
	      stderr);
	exit(1);
}


int main(int argc, char **argv)
{
	int	term_width = 80;
	int	c, i;
	int	num_cols = -1;
	char	*buf, *tmp;
	char	*remove=NULL;
	void	(*pwgen)(char *inbuf, int size, int pw_flags, char *remove);

	pwgen = pw_phonemes;
	pw_number = pw_random_number;
	if (isatty(1))
		do_columns = 1;
	pwgen_flags |= PW_DIGITS | PW_UPPERS;

	while (1) {
#ifdef HAVE_GETOPT_LONG
		c = getopt_long(argc, argv, pw_options, pwgen_options, 0);
#else
		c = getopt(argc, argv, pw_options);
#endif
		if (c == -1)
			break;
		switch (c) {
		case '0':
			pwgen_flags &= ~PW_DIGITS;
			break;
		case 'A':
			pwgen_flags &= ~PW_UPPERS;
			break;
		case 'a':
			break;
		case 'B':
			pwgen_flags |= PW_AMBIGUOUS;
			break;
		case 'c':
			pwgen_flags |= PW_UPPERS;
			break;
		case 'n':
			pwgen_flags |= PW_DIGITS;
			break;
		case 'N':
			num_pw = strtol(optarg, &tmp, 0);
			if (*tmp) {
				fprintf(stderr,
					"Invalid number of passwords: %s\n",
					optarg);
				exit(1);
			}
			break;
		case 's':
			pwgen = pw_rand;
			break;
		case 'C':
			do_columns = 1;
			break;
		case '1':
			do_columns = 0;
			break;
		case 'H': 
			pw_sha1_init(optarg);
			pw_number = pw_sha1_number;
			break;
		case 'y':
			pwgen_flags |= PW_SYMBOLS;
			break;
		case 'v':
			pwgen = pw_rand;
			pwgen_flags |= PW_NO_VOWELS;
			break;
		case 'r':
			remove = strdup(optarg);
			pwgen = pw_rand;
			break;
		case 'e':
			do_crypts = 1;
			char *temp_salt = getenv("SALTPREFIX");
			if (temp_salt) {
				salt_prefix=temp_salt;
			}
			break;
		case 'h':
		case '?':
			usage();
			break;
		}
	}
	if (optind < argc) {
		pw_length = strtol(argv[optind], &tmp, 0);
		if (pw_length < 5)
			pwgen = pw_rand;
		if (pwgen != pw_rand) {
			if (pw_length <= 2)
				pwgen_flags &= ~PW_UPPERS;
			if (pw_length <= 1)
				pwgen_flags &= ~PW_DIGITS;
		}
		if (*tmp) {
			fprintf(stderr, "Invalid password length: %s\n",
				argv[optind]);
			exit(1);
		}
		optind++;
	}

	if (optind < argc) {
		num_pw = strtol(argv[optind], &tmp, 0);
		if (*tmp) {
			fprintf(stderr, "Invalid number of passwords: %s\n",
				argv[optind]);
			exit(1);
		}
	}
	
	if (do_columns) {
		num_cols = term_width / (pw_length+1);
		if (num_cols == 0)
			num_cols = 1;
	}
	if (num_pw < 0)
		num_pw = do_columns ? num_cols * 20 : 1;
	
	buf = malloc(pw_length+1);
	if (!buf) {
		fprintf(stderr, "Couldn't malloc password buffer.\n");
		exit(1);
	}
	for (i=0; i < num_pw; i++) {
		pwgen(buf, pw_length, pwgen_flags, remove);
		if (!do_columns || ((i % num_cols) == (num_cols-1)) ||
		    (i == (num_pw - 1))) {

			printf("%s", buf);
			if (do_crypts) {
				printf(" %s", crypt(buf, crypt_gensalt(salt_prefix, 0, NULL,0)));
			}
			printf("\n");
		}
		else {
			printf("%s ", buf);
			if (do_crypts) {
				printf("%s ", crypt(buf, crypt_gensalt(salt_prefix, 0, NULL,0)));
			}
		}
	}
	free(buf);
	return 0;
}
